Docs reorg (#3039)
edgedb/edgedb: docs/edgeql/sets.rst

.. _ref_eql_sets:

Sets
====

- :ref:`Everything is a set <ref_eql_everything_is_a_set>`
- :ref:`Declaring sets <ref_eql_set_constructor>`
- :ref:`Literals are singletons <ref_eql_set_literals_are_singletons>`
- :ref:`Empty sets <ref_eql_empty_sets>`
- :ref:`Set references <ref_eql_set_references>`
- :ref:`Multisets <ref_eql_set_distinct>`
- :ref:`Checking membership <ref_eql_set_in>`
- :ref:`Merging <ref_eql_set_union>`
- :ref:`Coalescing <ref_eql_set_coalesce>`
- :ref:`Inheritance <ref_eql_set_type_filter>`
- :ref:`Aggregate vs element-wise operations <ref_eql_set_aggregate>`
- :ref:`Conversion to/from arrays <ref_eql_set_array_conversion>`

.. _ref_eql_everything_is_a_set:

Everything is a set
-------------------

> All values in EdgeQL is actually **sets**: a collection of values of a given

Reviewer: ```suggestion
All values in EdgeQL are actually **sets**: a collection of values of a given
```

expression. For instance, ``"hello world"`` is *exactly equivalent* to ``
{"hello world"}``.

.. code-block:: edgeql-repl

  db> select {"hello world"};
  {"hello world"}
  db> select "hello world" = {"hello world"};
  {true}

You can retrieve the cardinality of a set with the :eql:func:`count` function.

.. code-block:: edgeql-repl

  db> select count('aaa');
  {1}
  db> select count({'aaa', 'bbb'});
  {2}

.. _ref_eql_empty_sets:

Empty sets
----------

The reason EdgeQL introduced the concept of *sets* is to eliminate the concept
of ``NULL``. In SQL databases ``NULL`` is a special value denoting the absence
of data; in EdgeDB the absence of data is just an empty set.

.. note::

  Why is the existence of NULL a problem? Put simply, it's an edge case that
  permeates all of SQL and is often handled inconsistly in different
  circumstances. A number of specific inconsistencies are documented in detail
  in the `We Can Do Better Than SQL
  </blog/we-can-do-better-than-sql#null-a-bag-of-surprises>`_ post on the
  EdgeDB blog. For broader context, see Tony Hoare's talk
  `"The Billion Dollar Mistake" <https://bit.ly/3H238oG>`_.

> Declaring empty sets isn't as simply as ``{}``. In EdgeQL, all expressions are

Reviewer 2: ```suggestion
Declaring empty sets isn't as simple as ``{}``. In EdgeQL, all expressions are
```

*strongly typed*, including empty sets. With nonempty sets (like ``{1, 2, 3}``)
, the type is inferred from the set's contents (``int64``). But with empty sets
this isn't possible, so an *explicitly cast* is required.

.. code-block:: edgeql-repl

  db> select {};
  error: QueryError: expression returns value of indeterminate type
    ┌─ query:1:8
    │
  1 │ select {};
    │        ^^ Consider using an explicit type cast.

  db> select <int64>{};
  {}
  db> select <str>{};
  {}
  db> select count(<str>{});
  {0}

> .. code-block:: edgeql-repl
> 
>   db> select {};
>   error: QueryError: expression returns value of indeterminate type
>     ┌─ query:1:8

Reviewer 2: Repeated fragment of previous codeblock
```suggestion
```

.. code-block:: edgeql-repl

  db> select count({'aaa', 'bbb', 'ccc'})
  {2}
  db> select sum({1, 2, 3});
  {6}
  db> select min({1, 2, 3});
  {-3}

When an *element-wise* operation accepts two inputs, the operation is applied
*pair-wise*; in other words, the operation is applied to the *cartesian
product* of the inputs.

.. code-block:: edgeql-repl

  db> select {'aaa', 'bbb'} ++ {'ccc', 'ddd'}
  {'aaaccc', 'aaaddd', 'bbbccc', 'bbbddd'}

Accordingly, operations involving an empty set typically return an empty set
(though certain operations like :eql:func:`count` are able to operate on empty
sets).

.. code-block:: edgeql-repl

  db> select <str>{} ++ 'ccc'
  {}

.. _ref_eql_set_array_conversion:

Conversion to/from arrays
-------------------------

Both arrays and sets are collections of values that share a type. EdgeQL
provides ways to convert one into the other.

.. code-block:: edgeql-repl

  db> select array_unpack([1,2,3]);
  {1, 2, 3}
  db> select array_agg({1,2,3});
  {[1, 2, 3]}

You can perform many of the same operations on either sets or arrays. For
instance, the :eql:func:`count` operation on sets is analogous to the

> :eql:func:`len` operation on arrays. Whether you prefer to deal primary with

Reviewer 2: ```suggestion
:eql:func:`len` operation on arrays. Whether you prefer to deal primarily with
```

Reviewer 2: Also I think arrays have a definite order, but sets don't (unless they've been `order by`ed).

<ref_eql_select_backlinks>`.

.. _ref_eql_set_aggregate:

Aggregate vs element-wise operations
------------------------------------

EdgeQL provides a large library of built-in functions and operators for
handling data structures. Each functions and operators is either *aggregate* or
*element-wise*.

Element-wise operations are applied on *each element* of a set.

.. code-block:: edgeql-repl

  db> select str_upper({'aaa', 'bbb'})
  {'AAA', 'BBB'}
  db> select {1, 2, 3} ^ 2;
  {1, 4, 9}
  db> select str_split({"hello world", "hi again"}, " ");
  {["hello", "world"], ["hi", "again"]}

By contrast, *aggregate* operations accept a set with arbitrary cardinality and
return a *singleton* (or perhaps an empty set if the input was also empty).

.. code-block:: edgeql-repl

  db> select count({'aaa', 'bbb', 'ccc'})
  {2}
  db> select sum({1, 2, 3});
  {6}
  db> select min({1, 2, 3});
  {-3}

When an *element-wise* operation accepts two inputs, the operation is applied
*pair-wise*; in other words, the operation is applied to the *cartesian
product* of the inputs.

.. code-block:: edgeql-repl

  db> select {'aaa', 'bbb'} ++ {'ccc', 'ddd'}
  {'aaaccc', 'aaaddd', 'bbbccc', 'bbbddd'}

> Accordingly, operations involving an empty set typically return an empty set
> (though certain operations like :eql:func:`count` are able to operate on empty
> sets).

Reviewer 3: ```suggestion
Accordingly, operations involving an empty set typically return an empty set
(though aggregate functions like :eql:func:`count` operate on sets as a whole
and can handle empty sets).
```
